pkg/kubelet/volumemanager/volume_manager.go

```diff
@@ -332,7 +332,7 @@ func (vm *volumeManager) GetVolumesInUse() []v1.UniqueVolumeName {
 	}
 
 	for _, volume := range allAttachedVolumes {
-		if volume.PluginIsAttachable {
+		if volume.PluginIsAttachable || volume.DeviceMountState != operationexecutor.DeviceNotMounted {
```

Comment:
What is DeviceMountState for volume that has no attacher or device mounter?

Reply:
I checked the behavior for a driver which does not support node stage/unstage, only node publish/unpublish. (e.g. csi filestore driver with only node publish support).

When the volume is mounted, it still shows up in the attached volumes map, with DeviceMountState == DeviceGloballyMounted. Thus even for such drivers the node.Status.VolumesInUse reports the volume.
e.g
mount device succeeds (node stage skipped):
```
Feb 17 04:10:43 e2e-test-saikatroyc-minion-group-s7lb kubelet[1487]: I0217 04:10:43.328968    1487 csi_attacher.go:319] kubernetes.io/csi: attacher.MountDevice STAGE_UNSTAGE_VOLUME capability not set. Skipping MountDevice...
Feb 17 04:10:43 e2e-test-saikatroyc-minion-group-s7lb kubelet[1487]: I0217 04:10:43.329004    1487 operation_generator.go:594] MountVolume.MountDevice succeeded for volume "pvc-d3f3bf28-3fd9-407d-b16e-452b10ac89ce" (UniqueName: "kubernetes.io/csi/filestore.csi.storage.gke.io^modeInstance/us-central1-b/pvc-d3f3bf28-3fd9-407d-b16e-452b10ac89ce/vol1") pod "web-server" (UID: "1e6e315b-02c2-463b-8a31-001ebf82d95f") device mount path "/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-d3f3bf28-3fd9-407d-b16e-452b10ac89ce/globalmount"
```
Node publish is triggered and succeeds
```
Feb 17 04:10:43 e2e-test-saikatroyc-minion-group-s7lb kubelet[1487]: I0217 04:10:43.334757    1487 csi_mounter.go:212] kubernetes.io/csi: created target path successfully [/var/lib/kubelet/pods/1e6e315b-02c2-463b-8a31-001ebf82d95f/volumes/kubernetes.io~csi/pvc-d3f3bf28-3fd9-407d-b16e-452b10ac89ce]

Feb 17 04:10:43 e2e-test-saikatroyc-minion-group-s7lb kubelet[1487]: I0217 04:10:43.334787    1487 csi_mounter.go:321] kubernetes.io/csi: CSIDriver "filestore.csi.storage.gke.io" requires pod information
Feb 17 04:10:43 e2e-test-saikatroyc-minion-group-s7lb kubelet[1487]: I0217 04:10:43.334803    1487 csi_client.go:206] kubernetes.io/csi: calling NodePublishVolume rpc [volid=modeInstance/us-central1-b/pvc-d3f3bf28-3fd9-407d-b16e-452b10ac89ce/vol1,target_path=/var/lib/kubelet/pods/1e6e315b-02c2-463b-8a31-001ebf82d95f/volumes/kubernetes.io~csi/pvc-d3f3bf28-3fd9-407d-b16e-452b10ac89ce/mount]

Feb 17 04:10:47 e2e-test-saikatroyc-minion-group-s7lb kubelet[1487]: I0217 04:10:47.009528    1487 operation_generator.go:672] MountVolume.SetUp succeeded for volume "pvc-d3f3bf28-3fd9-407d-b16e-452b10ac89ce" (UniqueName: "kubern
etes.io/csi/filestore.csi.storage.gke.io^modeInstance/us-central1-b/pvc-d3f3bf28-3fd9-407d-b16e-452b10ac89ce/vol1") pod "web-server" (UID: "1e6e315b-02c2-463b-8a31-001ebf82d95f")
```

node status reports the volumes in use, and the device state is DeviceGloballyMounted.
```
Feb 17 04:18:18 e2e-test-saikatroyc-minion-group-s7lb kubelet[1487]: I0217 04:18:18.306430    1487 volume_manager.go:336] volume.VolumeName kubernetes.io/csi/filestore.csi.storage.gke.io^modeInstance/us-central1-b/pvc-d3f3bf28-3fd9-407d-b16e-452b10ac89ce/vol1, mount state: %!(EXTRA operationexecutor.DeviceMountState=DeviceGloballyMounted)
```

When the pod is deleted, and volume is unmounted (node unpublish)
```
Feb 17 04:19:13 e2e-test-saikatroyc-minion-group-s7lb kubelet[1487]: I0217 04:19:13.245504    1487 operation_generator.go:797] UnmountVolume.TearDown succeeded for volume "kubernetes.io/csi/filestore.csi.storage.gke.io^modeInstan
ce/us-central1-b/pvc-d3f3bf28-3fd9-407d-b16e-452b10ac89ce/vol1" (OuterVolumeSpecName: "mypvc") pod "1e6e315b-02c2-463b-8a31-001ebf82d95f" (UID: "1e6e315b-02c2-463b-8a31-001ebf82d95f"). InnerVolumeSpecName "pvc-d3f3bf28-3fd9-407d-
b16e-452b10ac89ce". PluginName "kubernetes.io/csi", VolumeGidValue ""
```

then volume removed from the attached volumes map (node unstage call skipped)
```
Feb 17 04:19:13 e2e-test-saikatroyc-minion-group-s7lb kubelet[1487]: I0217 04:19:13.274691    1487 csi_attacher.go:526] kubernetes.io/csi: attacher.UnmountDevice STAGE_UNSTAGE_VOLUME capability not set. Skipping UnmountDevice...

Feb 17 04:19:13 e2e-test-saikatroyc-minion-group-s7lb kubelet[1487]: I0217 04:19:13.275877    1487 operation_generator.go:891] UnmountDevice succeeded for volume "pvc-d3f3bf28-3fd9-407d-b16e-452b10ac89ce" %!(EXTRA string=UnmountD
evice succeeded for volume "pvc-d3f3bf28-3fd9-407d-b16e-452b10ac89ce" (UniqueName: "kubernetes.io/csi/filestore.csi.storage.gke.io^modeInstance/us-central1-b/pvc-d3f3bf28-3fd9-407d-b16e-452b10ac89ce/vol1") on node "e2e-test-saikatroyc-minion-group-s7lb" )
```

[Note: In this case, when the pod is still referencing the volume, PVC protection controller would block the PVC deletion anyway. When pod is deleted, and volume is unpublished, node unstage is no-op, the node status will no longer report the volume in volumesInUse. So we should be good with volume deletion workflow]

Reply:
so for those volume only supports publish/unpublish, do we need to report them in node status? If it is not needed, it is better to not include them to reduce the overhead.

Reply:
@jingxu97 we report volumesInUse looking at the asw.attachedVolumes map. Even for volumes which only supports node publish/unpublish, the volume is present in the asw.attachedVolumes map. Its just that the node stage/unstage calls are skipped by the [csi_attacher](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/csi/csi_attacher.go#L593). At the end of node unpublish and node unstage(which is again a no-op), the volume is removed from the asw map and volume will no longer be reported in volumesInUse. I didnt see a fine grained way to detect volumes which only support node unpublish.     Let's discuss more on this offline. The current code logic does not impact delete workflows for volumes which only supports node publish/unpublish. But lets discuss offline if there may be any side effects that I may have missed.

Reply:
Thoughts @msau42 on the comment https://github.com/kubernetes/kubernetes/pull/98890#discussion_r584004875 ?

Reply:
Currently we do not query plugin capabilities during plugin initialization and hence we don't know if a CSI plugin supports stage/unstage when initializing the device mounter. We could theoretically fix this by putting something similar to attachable property in `CSIDriver` object.

Reply:
@gnufied that seems like a good idea, but given that this itself is an involved fix to add a new field to the csidriver object, perhaps this should be tracked as a separate issue, and then we can enhance this current logic to take into account csi node stage capability.

Other than that, do you forsee any issue with the current logic, where we look at asw.attachedVolumes (even though the node stage/unstage is a no-op).

Reply:
@saikat-royc can you clarify this comment:
> does not impact delete workflows for volumes which only supports node publish/unpublish

Reply:
@msau42  I mean I tried out this fix with a driver with support for only node publish/unpublish support (i.e node stage/unstage), e.g. filestore driver (with some tweaks to remove node unstage/stage capability), and the fix works as expected, and volumes deleted successfully. volumes were reported in use when the volume is present in asw. attachedVolumes(), and not reported when its not present in asw.attachedVolumes()

high level sequence of calls:

1. node stage (no op)
2. volume populated in asw.attachedVolumes
3. node publish (suceeds)

in teardown,
1. node unpublish(succeeds)
2. node unstage (no op)
3. volume removed from asw.attachedVolumes
4. volume no longer reported in node status volumes in use.
5. delete proceeds as expected from the provisioner

Reply:
> Other than that, do you forsee any issue with the current logic, where we look at asw.attachedVolumes (as the node stage/unstage is a no-op for certain csi drivers which dont support it).

I commented below, I think we should be okay. But this is very subtle code change and it is easy to forget the full context of everything that is going on here and hence a detailed comment and some tests would be appreciated.

Reply:
Ack @gnufied . I agree this piece of logic was not immediately clear to me, I had to experiment and look at logs to understand what's happening

Reply:
Added some more details in comments about the csi volume behavior